text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+        /// Type-based `TypeTag` for `&'a T` types.
+        pub struct Ref<T: ?Sized + 'static>(PhantomData<T>);
+
+        impl<'a, T: ?Sized + 'static> TypeTag<'a> for Ref<T> {
+            type Type = &'a T;
+        }
+
+        /// Type-based `TypeTag` for `&'a mut T` types.
+        pub struct RefMut<T: ?Sized + 'static>(PhantomData<T>);
+
+        impl<'a, T: ?Sized + 'static> TypeTag<'a> for RefMut<T> {
+            type Type = &'a mut T;
+        }
+
+        /// Type-based `TypeTag` for static `T` types.
+        pub struct Value<T: 'static>(PhantomData<T>);
+
+        impl<'a, T: 'static> TypeTag<'a> for Value<T> {
+            type Type = T;
+        }
+
+        /// Tag combinator to wrap the given tag's value in an `Option<T>`
+        pub struct Optional<I>(PhantomData<I>);
+
+        impl<'a, I: TypeTag<'a>> TypeTag<'a> for Optional<I> {
+            type Type = Option<I::Type>;
+        }
+    }
+}
+```
+
+For intermediate libraries, `Value` serves the common case of providing a new or temporary value (for example, computing a `String` message from multiple fields), `Ref` and `RefMut` cover the common case of providing a reference to a field of the providing object (as in `Error::context_ref` providing a reference to `self.backtrace` in the motivating examples), where the field does not contain non-static references. `Optional` is used in the implementation of `provide_any`, it is public since it seems like it could be generally useful.
+
+### Requisition
+
+```rust
+impl<'a, 'b> Requisition<'a, 'b> {
+    /// Provide a value or other type with only static lifetimes.
+    pub fn provide_value<T, F>(&mut self, f: F) -> &mut Self
+    where
+        T: 'static,
+        F: FnOnce() -> T,
+    { ... }
+
+    /// Provide a reference, note that the referee type must be bounded by `'static`, but may be unsized.
+    pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self { ... }
+
+    /// Provide a value with the given `TypeTag`.
+    pub fn provide<I: TypeTag<'a>>(&mut self, value: I::Type) -> &mut Self { ... }
+
+    /// Provide a value with the given `TypeTag`, using a closure to prevent unnecessary work.
+    pub fn provide_with<I, F>(&mut self, f: F) -> &mut Self
+    where
+        I: TypeTag<'a>,
+        F: FnOnce() -> I::Type,
+    { ... }
+
+}
+
+```
+
+`Requisition` is an object for provider types to provide data to be accessed. It is required because there must be somewhere for the data (or a reference to it) to exist.
+
+`provide_value` and `provide_ref` are convenience methods for the common cases of providing a temporary value and providing a reference to a field of `self`, respectively. `provide_value` takes a function to avoid unnecessary work when querying for data of a different type; `provide_ref` does not use a function since creating a reference is typically cheap.
+
+`provide` and `provide_with` offer full generality, but require the explicit use of type tags.
```

Comment:
As use of this interface grows, use of `provide_with` to defer computation such as decompression are likely to come up. I'd like to see an example use of it included.